<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="igvShiny">
<title>igvShiny: a wrapper of IGV in the Shiny apps • igvShiny</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="igvShiny: a wrapper of IGV in the Shiny apps">
<meta property="og:description" content="igvShiny">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">igvShiny</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/igvShiny.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/gladkia/igvShiny/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>igvShiny: a wrapper of IGV in the Shiny apps</h1>
                        <h4 data-toc-skip class="author">Arkadiusz
Gladki</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/gladkia/igvShiny/blob/HEAD/vignettes/igvShiny.Rmd" class="external-link"><code>vignettes/igvShiny.Rmd</code></a></small>
      <div class="d-none name"><code>igvShiny.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="igvshiny">igvShiny<a class="anchor" aria-label="anchor" href="#igvshiny"></a>
</h2>
<p>The <code>igvShiny</code> package is a wrapper of Integrative
Genomics Viewer (IGV). It comprises an htmlwidget version of IGV. It can
be used as a module in Shiny apps.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"igvShiny"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-the-package">Loading the package<a class="anchor" aria-label="anchor" href="#loading-the-package"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-minimal-shiny-app">Running minimal Shiny app<a class="anchor" aria-label="anchor" href="#running-minimal-shiny-app"></a>
</h2>
<p>Running the minimal Shiny app with igvShiny is as simple as:</p>
<ul>
<li>providing genome details via
<code>parseAndValidateGenomeSpec</code>
</li>
<li>using <code>igvShinyOutput</code> as the UI function</li>
<li>using <code>renderIgvShiny</code> as the server function</li>
</ul>
<p>The he minimal Shiny app make look like:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span><span class="va">options</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseAndValidateGenomeSpec.html">parseAndValidateGenomeSpec</a></span><span class="op">(</span>genomeName<span class="op">=</span><span class="st">"hg38"</span>,  initialLocus<span class="op">=</span><span class="st">"NDUFS2"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ui</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyUI.html" class="external-link">shinyUI</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidPage</a></span><span class="op">(</span><span class="fu"><a href="../reference/igvShinyOutput.html">igvShinyOutput</a></span><span class="op">(</span><span class="st">'igvShiny'</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">server</span> <span class="op">&lt;-</span></span>
<span>        <span class="kw">function</span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span>, <span class="va">session</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">output</span><span class="op">$</span><span class="va">igvShiny</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderIgvShiny.html">renderIgvShiny</a></span><span class="op">(</span><span class="op">{</span></span>
<span>                <span class="fu"><a href="../reference/igvShiny.html">igvShiny</a></span><span class="op">(</span><span class="va">options</span><span class="op">)</span></span>
<span>            <span class="op">}</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html" class="external-link">shinyApp</a></span><span class="op">(</span>ui <span class="op">=</span> <span class="va">ui</span>, server <span class="op">=</span> <span class="va">server</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="providing-genome-details">Providing genome details<a class="anchor" aria-label="anchor" href="#providing-genome-details"></a>
</h2>
<p>Multiple genomes are currently supported by IGV: <a href="https://s3.amazonaws.com/igv.org.genomes/genomes.json" class="external-link">link</a>.
In igvShiny this set of genomes is called stock genomes. One can select
any stock genome easily by running
<code>parseAndValidateGenomeSpec</code> with single
<code>genomeName</code> value properly assigned. For example to use the
most popular mouse genome one need to run:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">igvShiny</span><span class="fu">::</span><span class="fu"><a href="../reference/parseAndValidateGenomeSpec.html">parseAndValidateGenomeSpec</a></span><span class="op">(</span><span class="st">"mm10"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $stockGenome</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dataMode</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $validated</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $genomeName</span></span>
<span><span class="co">#&gt; [1] "mm10"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $initialLocus</span></span>
<span><span class="co">#&gt; [1] "all"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fasta</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fastaIndex</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $annotation</span></span>
<span><span class="co">#&gt; [1] NA</span></span></code></pre></div>
<p>The list of available stock genomes in igvShiny can be found
with:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">igvShiny</span><span class="fu">::</span><span class="fu"><a href="../reference/get_css_genomes.html">get_css_genomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "hs1"             "chm13v1.1"       "hg38"            "hg38_1kg"       </span></span>
<span><span class="co">#&gt;  [5] "hg19"            "hg18"            "mm39"            "mm10"           </span></span>
<span><span class="co">#&gt;  [9] "mm9"             "rn7"             "rn6"             "gorGor6"        </span></span>
<span><span class="co">#&gt; [13] "gorGor4"         "panTro6"         "panTro5"         "panTro4"        </span></span>
<span><span class="co">#&gt; [17] "macFas5"         "GCA_011100615.1" "panPan2"         "canFam3"        </span></span>
<span><span class="co">#&gt; [21] "canFam4"         "canFam5"         "bosTau9"         "bosTau8"        </span></span>
<span><span class="co">#&gt; [25] "susScr11"        "galGal6"         "GCF_016699485.2" "danRer11"       </span></span>
<span><span class="co">#&gt; [29] "danRer10"        "ce11"            "dm6"             "dm3"            </span></span>
<span><span class="co">#&gt; [33] "dmel_r5.9"       "sacCer3"         "ASM294v2"        "ASM985889v3"    </span></span>
<span><span class="co">#&gt; [37] "tair10"          "GCA_003086295.2" "GCF_001433935.1" "NC_016856.1"    </span></span>
<span><span class="co">#&gt; [41] "GCA_000182895.1"</span></span></code></pre></div>
<p>See also demo app for stock genomes when one can select genome of
interest and familiarize with the basic functionalities provided via
igvShiny:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span><span class="va">demo_app_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span> <span class="st">"igvShiny"</span>, <span class="st">"demos"</span>, <span class="st">"stockGenomesDemo.R"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">demo_app_file</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span></code></pre></div>
<p>It’s also possible to use custom genomes (i.e. non-stock genomes). In
such cases one has to provide data for: <code>FASTA file</code>,
<code>FASTA index file</code> and <code>genome annotation file</code>.
The files can provided as local paths (<code>dataMode</code> =
<code>localFiles</code> or via URLs (<code>dataMode = http</code>). See
below the examples for both cases.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># defining custom genome with data provided via URLs</span></span>
<span><span class="va">base_url</span> <span class="op">&lt;-</span> <span class="st">"https://igv-data.systemsbiology.net/testFiles"</span></span>
<span><span class="va">title</span> <span class="op">&lt;-</span> <span class="st">"ribo remote"</span></span>
<span><span class="va">fasta_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base_url</span>, <span class="st">"ribosomal-RNA-gene.fasta"</span><span class="op">)</span></span>
<span><span class="va">fastaIndex_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base_url</span>, <span class="st">"ribosomal-RNA-gene.fasta.fai"</span><span class="op">)</span></span>
<span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/%s"</span>, <span class="va">base_url</span>, <span class="st">"ribosomal-RNA-gene.gff3"</span><span class="op">)</span></span>
<span><span class="va">locus</span> <span class="op">&lt;-</span> <span class="st">"U13369.1:7,276-8,225"</span></span>
<span><span class="va">genomeOptions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseAndValidateGenomeSpec.html">parseAndValidateGenomeSpec</a></span><span class="op">(</span></span>
<span>    genomeName <span class="op">=</span> <span class="va">title</span>,</span>
<span>    initialLocus <span class="op">=</span> <span class="va">locus</span>,</span>
<span>    stockGenome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    dataMode <span class="op">=</span> <span class="st">"http"</span>,</span>
<span>    fasta <span class="op">=</span> <span class="va">fasta_file</span>,</span>
<span>    fastaIndex <span class="op">=</span> <span class="va">fastaIndex_file</span>,</span>
<span>    genomeAnnotation <span class="op">=</span> <span class="va">annotation_file</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">genomeOptions</span></span>
<span><span class="co">#&gt; $stockGenome</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dataMode</span></span>
<span><span class="co">#&gt; [1] "http"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $validated</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $genomeName</span></span>
<span><span class="co">#&gt; [1] "ribo remote"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fasta</span></span>
<span><span class="co">#&gt; [1] "https://igv-data.systemsbiology.net/testFiles/ribosomal-RNA-gene.fasta"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fastaIndex</span></span>
<span><span class="co">#&gt; [1] "https://igv-data.systemsbiology.net/testFiles/ribosomal-RNA-gene.fasta.fai"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $initialLocus</span></span>
<span><span class="co">#&gt; [1] "U13369.1:7,276-8,225"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $annotation</span></span>
<span><span class="co">#&gt; [1] "https://igv-data.systemsbiology.net/testFiles/ribosomal-RNA-gene.gff3"</span></span>
<span></span>
<span><span class="co"># defining custom genome with data provided with local files</span></span>
<span><span class="va">data_directory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"igvShiny"</span>, <span class="st">"extdata"</span><span class="op">)</span></span>
<span><span class="va">fasta_file</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"ribosomal-RNA-gene.fasta"</span><span class="op">)</span></span>
<span><span class="va">fastaIndex_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"ribosomal-RNA-gene.fasta.fai"</span><span class="op">)</span></span>
<span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"ribosomal-RNA-gene.gff3"</span><span class="op">)</span></span>
<span><span class="va">genomeOptions2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseAndValidateGenomeSpec.html">parseAndValidateGenomeSpec</a></span><span class="op">(</span></span>
<span>    genomeName <span class="op">=</span> <span class="st">"ribo local"</span>,</span>
<span>    initialLocus <span class="op">=</span> <span class="st">"U13369.1:7,276-8,225"</span>,</span>
<span>    stockGenome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    dataMode <span class="op">=</span> <span class="st">"localFiles"</span>,</span>
<span>    fasta <span class="op">=</span> <span class="va">fasta_file</span>,</span>
<span>    fastaIndex <span class="op">=</span> <span class="va">fastaIndex_file</span>,</span>
<span>    genomeAnnotation <span class="op">=</span> <span class="va">annotation_file</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">genomeOptions2</span></span>
<span><span class="co">#&gt; $stockGenome</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dataMode</span></span>
<span><span class="co">#&gt; [1] "localFiles"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $validated</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $genomeName</span></span>
<span><span class="co">#&gt; [1] "ribo local"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fasta</span></span>
<span><span class="co">#&gt; [1] "/__w/_temp/Library/igvShiny/extdata/ribosomal-RNA-gene.fasta"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fastaIndex</span></span>
<span><span class="co">#&gt; [1] "/__w/_temp/Library/igvShiny/extdata/ribosomal-RNA-gene.fasta.fai"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $initialLocus</span></span>
<span><span class="co">#&gt; [1] "U13369.1:7,276-8,225"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $annotation</span></span>
<span><span class="co">#&gt; [1] "/__w/_temp/Library/igvShiny/extdata/ribosomal-RNA-gene.gff3"</span></span></code></pre></div>
<p>See also demo apps for custom genomes with data provided via
URLs:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span><span class="va">demo_app_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    package <span class="op">=</span> <span class="st">"igvShiny"</span>, </span>
<span>    <span class="st">"demos"</span>, </span>
<span>    <span class="st">"igvShinyDemo-customGenome-http.R"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">demo_app_file</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
<p>as well as data provided via local files:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span><span class="va">demo_app_file</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>        package <span class="op">=</span> <span class="st">"igvShiny"</span>, </span>
<span>        <span class="st">"demos"</span>, </span>
<span>        <span class="st">"igvShinyDemo-customGenome-localFiles.R"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">demo_app_file</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="main-functionalities">Main functionalities<a class="anchor" aria-label="anchor" href="#main-functionalities"></a>
</h2>
<p>In principle igvShiny provides the same functionalities that one can
find in igv.js. In summary following actions are currently possible
(wrapped as the R helpers):</p>
<ul>
<li>load tracks with the following formats/data types:
<ul>
<li>Bed</li>
<li>BedGraph</li>
<li>Seg</li>
<li>GWAS</li>
<li>Bam</li>
<li>Cram</li>
<li>Vcf</li>
<li>GFF3</li>
</ul>
</li>
<li>showGenomicRegion (zoom in or out to show the nominated region, by
chromosome locus or gene symbol)</li>
<li>getGenomicRegion (return the current IGV region)</li>
<li>removeUserAddedTracks (remove tracks from the browser, added during
the session by the user)</li>
</ul>
<p>See also demo app to familiarize with the basic functionalities
provided by igvShiny:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gladkia/igvShiny" class="external-link">igvShiny</a></span><span class="op">)</span></span>
<span><span class="va">demo_app_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span> <span class="st">"igvShiny"</span>, <span class="st">"demos"</span>, <span class="st">"igvShinyDemo.R"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">demo_app_file</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] igvShiny_0.99.4      shiny_1.8.0          GenomicRanges_1.54.1</span></span>
<span><span class="co">#&gt; [4] GenomeInfoDb_1.38.6  IRanges_2.36.0       S4Vectors_0.40.2    </span></span>
<span><span class="co">#&gt; [7] BiocGenerics_0.48.1  BiocStyle_2.30.0    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] xfun_0.42               bslib_0.6.1             htmlwidgets_1.6.4      </span></span>
<span><span class="co">#&gt;  [4] vctrs_0.6.5             tools_4.3.2             bitops_1.0-7           </span></span>
<span><span class="co">#&gt;  [7] curl_5.2.0              cluster_2.1.6           checkmate_2.3.1        </span></span>
<span><span class="co">#&gt; [10] randomcoloR_1.1.0.1     desc_1.4.3              lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt; [13] GenomeInfoDbData_1.2.11 compiler_4.3.2          stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [16] textshaping_0.3.7       munsell_0.5.0           httpuv_1.6.14          </span></span>
<span><span class="co">#&gt; [19] htmltools_0.5.7         sass_0.4.8              RCurl_1.98-1.14        </span></span>
<span><span class="co">#&gt; [22] yaml_2.3.8              later_1.3.2             pkgdown_2.0.7          </span></span>
<span><span class="co">#&gt; [25] jquerylib_0.1.4         ellipsis_0.3.2          cachem_1.0.8           </span></span>
<span><span class="co">#&gt; [28] mime_0.12               digest_0.6.34           Rtsne_0.17             </span></span>
<span><span class="co">#&gt; [31] stringi_1.8.3           purrr_1.0.2             bookdown_0.37          </span></span>
<span><span class="co">#&gt; [34] fastmap_1.1.1           colorspace_2.1-0        cli_3.6.2              </span></span>
<span><span class="co">#&gt; [37] magrittr_2.0.3          scales_1.3.0            promises_1.2.1         </span></span>
<span><span class="co">#&gt; [40] backports_1.4.1         rmarkdown_2.25          lambda.r_1.2.4         </span></span>
<span><span class="co">#&gt; [43] XVector_0.42.0          httr_1.4.7              futile.logger_1.4.3    </span></span>
<span><span class="co">#&gt; [46] ragg_1.2.7              memoise_2.0.1           evaluate_0.23          </span></span>
<span><span class="co">#&gt; [49] knitr_1.45              V8_4.4.2                rlang_1.1.3            </span></span>
<span><span class="co">#&gt; [52] futile.options_1.0.1    Rcpp_1.0.12             xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [55] glue_1.7.0              BiocManager_1.30.22     formatR_1.14           </span></span>
<span><span class="co">#&gt; [58] jsonlite_1.8.8          R6_2.5.1                systemfonts_1.0.5      </span></span>
<span><span class="co">#&gt; [61] fs_1.6.3                zlibbioc_1.48.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paul Shannon, Arkadiusz Gladki, Karolina Scigocka.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
